src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp

```diff
@@ -104,321 +150,710 @@ enum Spells
+    DATA_FACEROLLER_ACHIEVEMENT = 1
 };
 
 class boss_eadric : public CreatureScript
 {
 public:
     boss_eadric() : CreatureScript("boss_eadric") { }
-    struct boss_eadricAI : public ScriptedAI
+
+    struct boss_eadricAI : public argent_challenge_baseAI
     {
-        boss_eadricAI(Creature* creature) : ScriptedAI(creature)
+        boss_eadricAI(Creature* creature) : argent_challenge_baseAI(creature) { }
+
+        void Initialize() override
         {
-            Initialize();
-            instance = creature->GetInstanceScript();
-            creature->SetReactState(REACT_PASSIVE);
-            creature->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE);
+            _faceroller = false;
         }
 
-        void Initialize()
+        void SetData(uint32 type, uint32 data) override
         {
-            uiVenganceTimer = 10000;
-            uiRadianceTimer = 16000;
-            uiHammerJusticeTimer = 25000;
-            uiResetTimer = 5000;
+            if (type != DATA_FACEROLLER_ACHIEVEMENT)
+                return;
 
-            bDone = false;
+            _faceroller = (data == 1);
         }
 
-        InstanceScript* instance;
-
-        uint32 uiVenganceTimer;
-        uint32 uiRadianceTimer;
-        uint32 uiHammerJusticeTimer;
-        uint32 uiResetTimer;
+        uint32 GetData(uint32 type) const override
+        {
+            if (type != DATA_FACEROLLER_ACHIEVEMENT)
+                return 0;
 
-        bool bDone;
+            return _faceroller ? 1 : 0;
+        }
 
-        void Reset() override
+        void OnReset() override
         {
             Initialize();
+            argent_challenge_baseAI::OnReset();
         }
 
-        void DamageTaken(Unit* /*done_by*/, uint32 &damage) override
+        void DamageTaken(Unit* /*done_by*/, uint32& damage) override
         {
+            if (instance->GetBossState(DATA_ARGENT_CHALLENGE) == DONE)
+            {
+                damage = 0;
+                return;
+            }
+
             if (damage >= me->GetHealth())
             {
                 damage = 0;
-                EnterEvadeMode();
-                me->SetFaction(FACTION_FRIENDLY);
-                bDone = true;
+                if (instance->GetBossState(DATA_ARGENT_CHALLENGE) != SPECIAL)
+                {
+                    instance->SetBossState(DATA_ARGENT_CHALLENGE, SPECIAL);
+                    events.Reset();
+                    Talk(SAY_DEFEATED_E);
+                }
             }
         }
 
-        void MovementInform(uint32 MovementType, uint32 /*Data*/) override
+        void JustEngagedWith(Unit* who) override
         {
-            if (MovementType != POINT_MOTION_TYPE)
+            if (instance->GetBossState(DATA_ARGENT_CHALLENGE) == SPECIAL)
                 return;
 
-            instance->SetData(BOSS_ARGENT_CHALLENGE_E, DONE);
+            DoCastAOE(SPELL_VENGEANCE);
+            events.ScheduleEvent(EVENT_RADIANCE, urand(7 * IN_MILLISECONDS, 15 * IN_MILLISECONDS));
```

Comment:
events.ScheduleEvent(EVENT_RADIANCE, 4ms, 15ms);

if im not wrong initialization like that alrdy exist in TC

Reply:
Not ms. Seconds